File: frontend/types/saved-contact.ts
// Typed contract for the Save Contact endpoints. Mirrors the BE
// `SavedContact` response schema (being built to match). A saved contact is
// a polymorphic pointer: `source` names the origin table (e.g.
// "discovered_email") and `contact_id` is that table's row id. The denormalized
// name/title/email/… fields are snapshotted by the BE at save time so the
// Saved Contacts list renders without re-joining the source table.
export interface SavedContact {
  id: number;
  source: string;
  contact_id: number;
  name: string | null;
  title: string | null;
  email: string | null;
  company: string | null;
  phone: string | null;
  linkedin_url: string | null;
  created_at: string;
}

File: backend/app/schemas/saved_contact.py
"""Pydantic shapes for /saved-contacts (Save Contact feature)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class SavedContactCreate(BaseModel):
    """Request body for ``POST /api/v1/saved-contacts``.

    Only the ``source`` discriminator and the ``contact_id`` reference are
    trusted from the client; the snapshot (name / title / email / company /
    phone / linkedin_url) is resolved SERVER-side from the source row so the
    client can't forge the saved contact's fields.
    """

    source: str
    contact_id: int


class SavedContactResponse(BaseModel):
    """One row in ``GET /api/v1/saved-contacts`` (and the ``POST`` echo)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    contact_id: int
    name: str | None = None
    title: str | None = None
    email: str | None = None
    company: str | None = None
    phone: str | None = None
    linkedin_url: str | None = None
    created_at: datetime


class SavedContactWithOwner(SavedContactResponse):
    """One row in ``GET /api/v1/integrations/saved-contacts``.

    A saved contact PLUS the id / name / email of the user who saved it, so the
    sibling CRM can attribute each contact to its owner across ALL users --
    unlike the owner-scoped ``SavedContactResponse`` served to end users. Owner
    name/email are typed optional to stay forgiving of sparse ``user`` rows even
    though the inner join guarantees an owner id.
    """

    saved_by_id: str
    saved_by_name: str | None = None
    saved_by_email: str | None = None
